BreathData.cpp

```diff
@@ -0,0 +1,146 @@
+  cycleTimer = millis();                   // reset the cycle timer at the start of inspiration
+
+  // record values from last breath
+  expDuration = cycleTimer - expTimer;        // measured duration of last expiration (EXP_STATE + PEEP_PAUSE + EXP_HOLD)
+  tidalVolumeExp = expFlowReader.getVolume(); // set current inspired volume as the measured inspiratory tidal volume for the last breath
+
+  // calculate actual minute volume from last breath
+  const float minuteVolume = inspFlowReader.getVolume() * CC_PER_MS_TO_LPM / cycleDuration;
+
+  // Update patient data on display to reflect values from last breath
+  display.writePeak(inspPressureReader.peak());        // measured pip cmH2O
+  display.writePlateau(inspPressureReader.plateau());  // measured plateau (only measured if HOLD_INSP_STATE is activated) cmH2O
+  display.writePeep(expPressureReader.peep());         // measured PEEP cmH2O
+  display.writeVolumeExp(expFlowReader.getVolume());   // measured expired volume
+  display.writeMinuteVolume(minuteVolume);             // measured minute volume
+  display.writeBPM(60000.0/cycleDuration);             // measured respiratory rate (seconds)
+  display.writeO2(O2);                                 // measured FIO2 concentration (@FutureWork: pending addition of O2 sensor)  
+
+  // close expiratory valve
+  expValve.close();
+
+  // Compute intervals at current settings
+  unsigned long targetCycleDuration = 60000UL / display.bpm(); // ms from start of cycle to end of inspiration
+  targetInspDuration = 105 * targetCycleDuration * display.inspPercent() / 10000; // allowing a bit more time to complete tidal volume inhailation
+  targetCycleEndTime = cycleTimer + targetCycleDuration;                          // target time for breath to end (for HOLD_EXP_STATE to end)
+  targetInspEndTime  = cycleTimer + targetInspDuration;                           // target time for INSP_STATE to end
+  targetExpDuration  = targetCycleDuration - targetInspDuration - MIN_PEEP_PAUSE; // target time for EXP_STATE to end
+  desiredInspFlow = display.volume() * CC_PER_MS_TO_LPM / targetInspDuration;     // desired inspiratory flowrate cc/ms
+
+  // begin PID control based on desired flow and reset tidal volume
+  inspValve.beginBreath(desiredInspFlow); 
+  inspFlowReader.resetVolume();           
+  cycleCount++;                           
+}
+
+
+/** 
+ * function to check if sensor readings are within acceptable ranges
+ * activate an alarm if they are not and deactivate the alarm once they are again
+ * 
+ * @param reading -- sensor reading
+ * @param compareValue -- value to compare reading to
+ * @param sensitivity -- the range of error
+ * @param highAlarmCode -- the high alarm code for a specific sensor
+ * @param lowAlarmCode -- the low alarm code for a specific sensor
+ */ 
+void BreathData::checkAlarmRangeWithUpdate(float reading, float &compareValue, float sensitivity, alarmCode highAlarmCode, alarmCode lowAlarmCode) { 
```

Comment:
Wouldn't this make more sense in the alarms module?

I'm a little uneasy about this function updating the `compareValue` this way.  An alternative way to write it, which would also avoid the need for the little function below, would be:
```
float BreathData::checkAlarmRange(float reading, float compareValue,
                                  float sensitivity, alarmCode highAlarmCode, alarmCode lowAlarmCode)
```
Which would return `compareValue` unchanged if an alarm is activated, otherwise it would return `reading`.  You would call it like this:
```
xyzLastReading = checkAlarmRange(xyzReading, xyzLastReading, ...);
```

Reply:
I like this suggestion!